waiter/src/waiter/shell_scheduler.clj

```diff
@@ -128,10 +128,11 @@
 
 (defn launch-instance
   "Launches a new process for the given service-id"
-  [service-id working-dir-base-path command environment port->reservation-atom port-range]
+  [service-id working-dir-base-path command environment num-ports port->reservation-atom port-range]
   (when-not command
     (throw (ex-info "The command to run was not supplied" {:service-id service-id})))
-  (let [port (reserve-port! port->reservation-atom port-range)
+  (let [reserved-ports (repeatedly num-ports #(reserve-port! port->reservation-atom port-range))
```

Comment:
Done, ended up using `atom`.

The following did not compile:
```
(defn reserve-ports!
  "Reserves num-ports available ports on the host, from the (optionally) provided range"
  [num-ports port->reservation-atom port-range]
  (try
    (loop [reserved-ports []
           remaining num-ports]
      (if (pos? remaining)
        (let [reserved-port-or-exception
              (try
                (reserve-port! port->reservation-atom port-range)
                (catch Throwable ex
                  (ex-info (str "Unable to reserve " num-ports " ports.")
                           {:num-successfully-reserved-before-error (count reserved-ports)
                            :reserved-ports reserved-ports}
                           ex)))]
          (if (instance? Throwable reserved-port-or-exception)
            reserved-port-or-exception
            (recur (conj reserved-ports reserved-port-or-exception) (dec remaining)))))
      reserved-ports)
    (catch Exception ex
      (throw ex))))
```

Error message:
```      
CompilerException java.lang.UnsupportedOperationException: Can only recur from tail position, compiling:(/Users/shamsimam/projects/github-projects/twosigma-waiter/waiter/src/waiter/shell_scheduler.clj:146:13) 
```